test/common/event/dispatcher_impl_test.cc

```diff
@@ -173,6 +173,49 @@ TEST_F(SchedulableCallbackImplTest, ScheduleChainingAndCancellation) {
   dispatcher_->run(Dispatcher::RunType::Block);
 }
 
+TEST_F(SchedulableCallbackImplTest, RescheduleNext) {
+  DispatcherImpl* dispatcher_impl = static_cast<DispatcherImpl*>(dispatcher_.get());
+  ReadyWatcher prepare_watcher;
+  evwatch_prepare_new(&dispatcher_impl->base(), onWatcherReady, &prepare_watcher);
+
+  ReadyWatcher watcher0;
+  createCallback([&]() {
+    watcher0.ready();
+    // Callback 1 was scheduled from the previous iteration, expect it to fire in the current
+    // iteration despite the attempt to reschedule.
+    callbacks_[1]->scheduleCallbackNextIteration();
+    // Callback 2 expected to execute next iteration because current because current called before
+    // next.
+    callbacks_[2]->scheduleCallbackCurrentIteration();
+    callbacks_[2]->scheduleCallbackNextIteration();
+    // Callback 3 expected to execute next iteration because next because next was called before
+    // current.
+    callbacks_[3]->scheduleCallbackNextIteration();
+    callbacks_[3]->scheduleCallbackCurrentIteration();
+  });
+
+  ReadyWatcher watcher1;
+  createCallback([&]() { watcher1.ready(); });
+  ReadyWatcher watcher2;
+  createCallback([&]() { watcher2.ready(); });
+  ReadyWatcher watcher3;
+  createCallback([&]() { watcher3.ready(); });
+
+  // Schedule callbacks 0 and 1 outside the loop, both will run in the same iteration of the event
```

Comment:
This is really a "gotcha", but probably there is nothing we can do as it is libevent behavior.

Reply:
This behavior is expected and fairly easy to reason about.  I mention it here since it relates to the operation done on callback 1 from callback 0, which is the main difference in behavior introduced by this PR: a next iteration callback scheduled in the previous iteration does not end up kicked down the road if the current iteration attempts to re-schedule it for the next iteration.